### PR TITLE
Replace usage of deque and map with vector in reference contracts

### PR DESCRIPTION
**Pull request by linh2931**
## Change Description

Resolve https://github.com/AntelopeIO/reference-contracts/issues/2

`CDT 3.0.0` introduced nested containers and map handing in ABI generation. It broke reference contracts' ABIs. See description and discussions in https://github.com/AntelopeIO/reference-contracts/issues/2 for details.

This PR replaces the usage of `map` and `deque` with `vector` and produces the same ABIs as prior to `CDT 3.0.0.`

## Deployment Changes
- [ ] Deployment Changes

## API Changes
- [ ] API Changes

## Documentation Additions
- [ ] Documentation Additions

### CODE REVIEW

File: contracts/eosio.system/src/rex.cpp [hunk 1]
```diff
@@ -664,7 +664,25 @@ namespace eosiosystem {
...
+                  } else if ( new_bucket_time == iter->first ) {
+                     *iter = pair_time_point_sec_int64 { new_bucket_time, new_bucket_rate };
+                     added = true;
+                     break;
+                  }
+                  ++iter;
+               }
+               if ( !added ) {
+                  return_buckets.emplace_back( pair_time_point_sec_int64 { new_bucket_time, new_bucket_rate });
+               }
```
Comment (arhag) [inline contracts/eosio.system/src/rex.cpp:685]: This does not appear to keep the buckets in `rb.return_buckets` sorted by `first`.

Instead replace with the follow which does maintain sort order and is simpler:
```cpp
auto iter = std::lower_bound(rb.return_buckets.begin(), rb.return_buckets.end());
if ((iter != return_buckets.end()) && (iter->first == new_bucket_time)) {
   iter->second = new_bucket_rate;
} else {
   rb.return_buckets.insert(iter, pair_time_point_sec_int64{new_bucket_time, new_bucket_rate});
}
```

File: contracts/eosio.system/src/rex.cpp [hunk 2]
```diff
@@ -677,14 +695,11 @@ namespace eosiosystem {
...
             while ( iter != return_buckets.end() && iter->first <= time_threshold ) {
-               auto next = iter;
-               ++next;
                const uint32_t overtime = get_elapsed_intervals( effective_time,
                                                                 iter->first + seconds(rex_return_pool::total_intervals * rex_return_pool::dist_interval) );
                surplus      += iter->second * overtime;
                expired_rate += iter->second;
-               return_buckets.erase(iter);
-               iter = next;
+               iter         =  return_buckets.erase(iter);
```
Comment (arhag) [inline contracts/eosio.system/src/rex.cpp:702]: Better yet delay the erase until the end:

```cpp
auto iter = return_buckets.begin();
for (; iter != return_buckets.end() && iter->first <= time_threshold; ++iter) {
   const uint32_t overtime = get_elapsed_intervals(effective_time,
                                                   iter->first + seconds(rex_return_pool::total_intervals * rex_return_pool::dist_interval));

   surplus      += iter->second * overtime;
   expired_rate += iter->second;
}
return_buckets.erase(return_buckets.begin(), iter);
```

**Review by arhag — CHANGES_REQUESTED** [review]

File: contracts/eosio.system/src/rex.cpp (hunk 1 above)
Comment (linh2931) [reply contracts/eosio.system/src/rex.cpp:685]: Thanks. Your way is cleaner and simpler. Will use it.

File: contracts/eosio.system/src/rex.cpp (hunk 2 above)
Comment (linh2931) [reply contracts/eosio.system/src/rex.cpp:702]: Will use it.

**Review by arhag — APPROVED** [review]

**Pull request merged**